File: core/src/test/java/com/spidey01/sxe/core/common/NotificationManagerTest.java
/*-
 * Copyright (c) 2013-current, Terry Mathew Poulin <BigBoss1964@gmail.com>
 *
 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the
 * use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *	1. The origin of this software must not be misrepresented; you must
 *	   not claim that you wrote the original software. If you use this
 *	   software in a product, an acknowledgment in the product
 *	   documentation would be appreciated but is not required.
 *
 *	2. Altered source versions must be plainly marked as such, and must
 *	   not be misrepresented as being the original software.
 *
 *	3. This notice may not be removed or altered from any source
 *	   distribution.
 */

package com.spidey01.sxe.core.common;

import com.spidey01.sxe.core.logging.Log;
import com.spidey01.sxe.core.testing.UnitTest;

import org.junit.*;

import java.util.EventObject;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;


public class NotificationManagerTest extends UnitTest {
    private static final String TAG = "NotificationManagerTest";


    /** Helper event class. */
    public class TestEvent extends EventObject {
        private final String mText;

        public TestEvent(Object source, String text) {
            super(source);
            mText = text;
        }

        public String toString() { return mText; }
    }


    /** Helper notification interface. */
    public interface TestListener {
        void onTest(TestEvent event);
    }


    /** Helper event manager. */
    public class TestEventManager extends NotificationManager<TestListener, TestEvent> {
        @Override
        protected void invoke(TestListener test, TestEvent event) {
            test.onTest(event);
        }
    }


    public class TestUser implements TestListener {
        public boolean called;
        public TestEvent lastTestEvent;

        public TestEventManager mTestEventManager = new TestEventManager();


        public void onTest(TestEvent event) {
            Log.i(TAG, "OnTest(", event.getClass().getName(), "=>", event, ")");
            called = true;
            lastTestEvent = event;
        }


        public void inject(String dummy) {
            mTestEventManager.notifyListeners(new TestEvent(this, dummy));
        }
    }


    @BeforeClass
    public static void beforeClass() {
        UnitTest.setup();
    }


    private TestUser mTestUser;


    @Before
    public void before() {
        Log.i(TAG, "before()");
        mTestUser = new TestUser();
    }


    @Test
    public void broadcast() {
        final String str = "foo";

        mTestUser.called = false;
        mTestUser.lastTestEvent = null;

        mTestUser.mTestEventManager.subscribe(mTestUser);


        assertNotCalled(mTestUser, str);
        mTestUser.inject(str);
        assertCalled(mTestUser, str);

        mTestUser.called = false;
        mTestUser.lastTestEvent = null;

        mTestUser.mTestEventManager.unsubscribe(mTestUser);

        assertNotCalled(mTestUser, str);
        mTestUser.inject(str);
        assertNotCalled(mTestUser, str);
    }


    @Test
    public void subscribed() {
        final String not_used_str = "foo";
        final String used_str = "bar";

        mTestUser.called = false;
        mTestUser.lastTestEvent = null;

        mTestUser.mTestEventManager.subscribe(used_str, mTestUser);


        assertNotCalled(mTestUser, not_used_str);
        mTestUser.inject(not_used_str);
        assertNotCalled(mTestUser, not_used_str);

        assertNotCalled(mTestUser, used_str);
        mTestUser.inject(used_str);
        assertCalled(mTestUser, used_str);

        mTestUser.called = false;
        mTestUser.lastTestEvent = null;

        mTestUser.mTestEventManager.unsubscribe(mTestUser);

        assertNotCalled(mTestUser, used_str);
        mTestUser.inject(used_str);
        assertNotCalled(mTestUser, used_str);
    }


    private void assertCalled(TestUser u, String s) {
        Assert.assertTrue(u.called);
        Assert.assertNotNull(u.lastTestEvent);
        Assert.assertEquals(s, u.lastTestEvent.toString());
    }


    private void assertNotCalled(TestUser u, String s) {
        Assert.assertFalse(mTestUser.called);
        Assert.assertNull(mTestUser.lastTestEvent);
    }

}

File: tests/core/common/NotificationManagerTest.cpp
/*-
 * Copyright (c) 2019-current, Terry Mathew Poulin <BigBoss1964@gmail.com>
 *
 * This software is provided 'as-is', without any express or implied warranty.
 * In no event will the authors be held liable for any damages arising from the
 * use of this software.
 *
 * Permission is granted to anyone to use this software for any purpose,
 * including commercial applications, and to alter it and redistribute it
 * freely, subject to the following restrictions:
 *
 *	1. The origin of this software must not be misrepresented; you must
 *	   not claim that you wrote the original software. If you use this
 *	   software in a product, an acknowledgment in the product
 *	   documentation would be appreciated but is not required.
 *
 *	2. Altered source versions must be plainly marked as such, and must
 *	   not be misrepresented as being the original software.
 *
 *	3. This notice may not be removed or altered from any source
 *	   distribution.
 */

#include "./NotificationManagerTest.hpp"

#include <sxe/core/sys/FileSystem.hpp>
#include <sxe/logging.hpp>
#include <sxe/stdheaders.hpp>

using std::string;
using sxe::core::common::NotificationManager;
using sxe::core::sys::FileSystem::path;

static const string TAG = "NotificationManagerTest";

CPPUNIT_TEST_SUITE_NAMED_REGISTRATION(NotificationManagerTest, TAG);
CPPUNIT_REGISTRY_ADD_TO_DEFAULT(TAG);


/** Helper event class.
 *
 * NotificationManager shouldn't need to care what this is.
 *
 * Just wrapping a string because K.I.S.S. More serious examples would be
 * something like an event from input device.
 */
class SXE_PRIVATE TestEvent
{
  private:
    string mText;

  public:
    TestEvent(const string& text)
        : mText(text)
    {
    }

    string toString() const
    {
        return mText;
    }

    void clear()
    {
        mText.clear();
    }

    operator string() const
    {
        return mText;
    }
};

/** Helper notification interface.
 *
 * NotificationManager just wants a invokable that takes the event.
 */
using TestListener = std::function<void(TestEvent)>;


/** Helper event manager. */
class SXE_PRIVATE TestEventManager : public NotificationManager<TestListener, TestEvent> {

    string_type asString(const TestEvent& event)
    {
        return event.toString();
    }

};


/** Helper user class.
 */
class SXE_PRIVATE TestUser
{
  public:

    bool called;
    TestEvent lastTestEvent;
    TestEventManager mTestEventManager;

    TestUser()
        : called(false)
        , lastTestEvent("")
        , mTestEventManager()
    {
        Log::xtrace(TAG, "TestUser ctor");
    }

    ~TestUser()
    {
        Log::xtrace(TAG, "TestUser dtor");
    }


    /** Callback to handle the event notification. */
    void onTest(TestEvent event)
    {
        string text = event.toString();

        Log::v(TAG, "TestUser::onTest(TestEvent => " + text + ")");

        called = true;
        lastTestEvent = event;
    }


    /** Injects a TestEvent to the manager. */
    void inject(const string& dummy)
    {
        mTestEventManager.notifyListeners(TestEvent(dummy));
    }
};


void NotificationManagerTest::broadcast()
{
    Log::xtrace(TAG, "broadcast()");

    static const string str = "foo";

    TestUser user;

    reset(user);

    Log::test(TAG, "Subscribing to broadcasts.");
    // Should work using a bind, etc.
    TestListener listener = std::bind(&TestUser::onTest, &user, std::placeholders::_1);
    auto id = user.mTestEventManager.subscribe(listener);

    assertNotCalled(user, str);
    Log::test(TAG, "Injecting str");
    user.inject(str);
    assertCalled(user, str);

    reset(user);

    Log::test(TAG, "Unsubscribing to broadcasts.");
    user.mTestEventManager.unsubscribe(id);

    assertNotCalled(user, str);
    user.inject(str);
    assertNotCalled(user, str);
}


void NotificationManagerTest::subscribe()
{
    Log::xtrace(TAG, "subscribe()");

    static const string not_used_str = "foo";
    static const string used_str = "bar";

    TestUser user;

    reset(user);

    // And with a lambda
    TestListener listener = [&user](auto str) -> void { user.onTest(TestEvent(str)); };

    Log::test(TAG, "Subscribing to used_str");
    auto id = user.mTestEventManager.subscribe(listener, used_str);

    Log::test(TAG, "Injecting a not_used_str.");
    assertNotCalled(user, not_used_str);
    user.inject(not_used_str);
    assertNotCalled(user, not_used_str);

    reset(user);

    Log::test(TAG, "Unsubscribing from all");
    user.mTestEventManager.unsubscribe(used_str, id);

    Log::test(TAG, "Injecting used_str after unsubscribe.");
    assertNotCalled(user, used_str);
    user.inject(used_str);
    assertNotCalled(user, used_str);
}


void NotificationManagerTest::reset(TestUser& u)
{
    Log::xtrace(TAG, "reset()");

    u.called = false;
    u.lastTestEvent.clear();
}


void NotificationManagerTest::assertCalled(const TestUser& u, const string& s)
{
    CPPUNIT_ASSERT_MESSAGE("TestUser::called is set when called.", 
                           u.called == true);
    CPPUNIT_ASSERT_MESSAGE("TestUser::lastTestEvent as a string should match the event string.",
                           s == u.lastTestEvent.toString());
}


void NotificationManagerTest::assertNotCalled(const TestUser& u, const string& s)
{
    CPPUNIT_ASSERT_MESSAGE("TestUser::called is unset until called.", 
                           u.called == false);
    CPPUNIT_ASSERT_MESSAGE("TestUser::lastTestEvent is nullptr until called.", 
                           u.lastTestEvent.toString().empty());
}
